lib/saml2.coffee

```diff
@@ -225,21 +235,46 @@ to_error = (err) ->
   return new Error(util.inspect err) unless err instanceof Error
   err
 
-# Takes in an xml @dom of an object containing an EncryptedAssertion and attempts to decrypt it using the @private_key.
-# @cb will be called with an error if the decryption fails, or the EncryptedAssertion cannot be found. If successful,
-# it will be called with the decrypted data as a string.
-decrypt_assertion = (dom, private_key, cb) ->
-  # This is needed because xmlenc sometimes throws an exception, and sometimes calls the passed in callback.
+# Takes in an XML @dom of an object containing an EncryptedAssertion and attempts to decrypt it
+# using the @private_keys in the given order.
+#
+# @cb will be called with an error if the decryption fails, or the EncryptedAssertion cannot be
+# found. If successful, it will be called with the decrypted data as a string.
+decrypt_assertion = (dom, private_keys, cb) ->
+  # This is needed because xmlenc sometimes throws an exception, and sometimes calls the passed-in
+  # callback.
   cb = _.wrap cb, (fn, err, args...) -> setTimeout (-> fn to_error(err), args...), 0
 
   try
     encrypted_assertion = dom.getElementsByTagNameNS(XMLNS.SAML, 'EncryptedAssertion')
-    return cb new Error("Expected 1 EncryptedAssertion; found #{encrypted_assertion.length}.") unless encrypted_assertion.length is 1
+    unless encrypted_assertion.length is 1
+      return cb new Error("Expected 1 EncryptedAssertion; found #{encrypted_assertion.length}.")
 
     encrypted_data = encrypted_assertion[0].getElementsByTagNameNS(XMLNS.XENC, 'EncryptedData')
-    return cb new Error("Expected 1 EncryptedData inside EncryptedAssertion; found #{encrypted_data.length}.") unless encrypted_data.length is 1
-
-    xmlenc.decrypt encrypted_data[0].toString(), (key: format_pem(private_key, 'PRIVATE KEY')), cb
+    unless encrypted_data.length is 1
+      return cb new Error("Expected 1 EncryptedData inside EncryptedAssertion; found #{encrypted_data.length}.")
+
+    encrypted_data = encrypted_data[0].toString()
+    decrypted_data = null
+    decryption_key_index = null
+    errors = []
+    async.eachOfSeries(
```

Comment:
While I give you that it looks strange, it's a pretty common coffeescript convention with async. Examples of this can be found [here](https://github.com/Clever/oauth/blob/master/authn/saml/index.coffee#L169-L192), [here](https://github.com/Clever/saml2/blob/master/lib/saml2.coffee#L357-L396) or [here](https://github.com/Clever/oauth/blob/master/authn/district_admin/index.coffee#L84-L104).